dump/match_infra.py

```diff
@@ -0,0 +1,287 @@
+            src = RedisSource()
+        else:
+            d_src = req.file
+            src = JsonSource()
+        
+        if not src.connect(d_src, req.ns):
+            template['error']  = error_dict["CONN_ERR"]
+            return self.__return_error(template)
+        verbose_print("MatchRequest Checks Passed")
+        all_matched_keys = src.getKeys(req.db, req.table, req.key_pattern)
+        if not all_matched_keys or len(all_matched_keys) == 0:
+            template['error'] = error_dict["INV_PTTRN"]
+            return self.__return_error(template)
+        
+        filtered_keys = self.__filter_out_keys(src, req, all_matched_keys)
+        verbose_print("Filtered Keys:" + str(filtered_keys))
+        if not filtered_keys:
+            template['error'] = error_dict["NO_ENTRIES"]
+            return self.__return_error(template)
+        return self.__fill(src, req, filtered_keys)
+    
+    def __ret_template(self):
+        return {"error" : "", "keys" : [], "return_values" : {}}
+    
+    
+    def __return_error(self, template):
+        verbose_print("MatchEngine: \n" + template['error'])
+        return template
+    
+    def __validate_request(self, req):
+        
+        if not isinstance(req, MatchRequest):
+            return error_dict["INV_REQ"]
+            
+        if not(req.db) and not(req.file):
+            return error_dict["NO_SRC"]
+        
+        if req.db and req.file:
+            return error_dict["SRC_VAGUE"]
+        
+        if not req.db and os.path.exists(req.file):
+            try:
+                with open(req.file) as f:
+                    json.load(f)
+            except ValueError as e:
+                return error_dict["INV_JSON"] 
+        elif not req.db:
+            return error_dict["NO_FILE"]
+        
+        if not(req.file) and req.db not in SonicDBConfig.getDbList():
+            return error_dict["INV_DB"]
+        
+        if not req.table:
+            return error_dict["NO_TABLE"]
+        
+        if not req.key_pattern:
+            return error_dict["NO_KEY"]
+        
+        if not isinstance(req.return_fields, list):
+            return error_dict["BAD_FORMAT_RE_FIELDS"]
+        
+        if not req.just_keys and len(req.return_fields) > 0:
+            return error_dict["JUST_KEYS_COMPAT"]
+        
+        if req.field and not req.value:
+            return error_dict["NO_VALUE"]
+        
+        return ""
+    
+    def __filter_out_keys(self, src, req, all_matched_keys):
+        if not (req.field):
+            return all_matched_keys
+        
+        filtered_keys = []
+        for key in all_matched_keys:
+            f_values = src.hget(req.db, key, req.field)
+            if "," in f_values and not req.match_entire_list: # Fields Containing Multiple Values
+                f_value = f_values.split(",")
+            else:
+                f_value = [f_values]
+            if req.value in f_value:
+                filtered_keys.append(key)
```

Comment:
Can user specify a custom filter callback (KeyFilterCallback)? I gues someday someone will want to filter keys by multiple field values or matching several values in a list or more complex scenarios.
You can have a set of filters provided by a framework, like this for example:
```has_value```
```has_value_in_list```
This introduces another indirection where MatchRequest - matches keys and a KeyFilterCallback is running against matched keys. We can have a call to discuss this.

Reply:
Will be handled in the next version of the feature